### ISSUE TITLE
Using more widely used symbol polyfill or means of detecting `$iterator$`

### ISSUE THREAD

**Issue by blesh** [paulmillr/es6-shim#427]
[Currently this library uses a library-specific string for iterators if Symbol.iterator does not exist](https://github.com/paulmillr/es6-shim/blob/3e46993a95b1867665166e0290e5cbaaf2ccb2f6/es6-shim.js#L266).

This has meant we need [a strange workaround for RxJS to access iterators](https://github.com/ReactiveX/rxjs/blob/0271fab2f435d21fc64187a39143237208a54163/src/symbol/iterator.ts#L18-L26) on types provided by es6-shim. This isn't ideal, because if you change your implementation of `Set` or `Map`, our workaround will break.

There are two fixes I'll propose for this, and neither are perfect:

1. Use a more common key for the iterator property
  - Basically use `"@@iterator"`.
2. Publish the `$iterator$` variable you're using somewhere globally so other libraries like RxJS can check for it and use it without workarounds.

**Comment by ljharb** [comment]
When we're in a browser that has a native property - like certain versions of Firefox with `"@@iterator"` - we use that. However, when we're not in a browser that has that, or `Symbol`, there's no global place to put it.

Where would you suggest?

Note that the "strange workaround" you're using is actually the perfect and correct approach to reliably get this value - I'm not sure why you'd need a different one.

**Comment by blesh** [comment]
> Note that the "strange workaround" you're using is actually the perfect and correct approach to reliably get this value - I'm not sure why you'd need a different one.

If you ever change your implementation of `Map` or `Set`, such that `Map.prototype[$yourIterator$] !== Map.prototype.entries`, then the workaround breaks.

**Comment by ljharb** [comment]
@blesh if we did that, we'd be violating the spec - thus, we will never do that.

**Comment by ljharb** [comment]
To clarify, the spec *requires* that `Map.prototype[Symbol.iterator] === Map.prototype.entries && Set.prototype[Symbol.iterator] === Set.prototype.values && Array.prototype[Symbol.iterator] === Array.prototype.values` be `true`.

**Comment by blesh** [comment]
Interesting. Thanks. I couldn't locate the spec on this.

It's still unfortunate that we have a special case for es6-shim due to a somewhat avant-garde symbol property name. (@@iterator would have been nice)

Thanks for your feedback.

**Comment by ljharb** [comment]
I explicitly chose not to use "@@iterator" to avoid people wrongly attempting to "polyfill" iterators based on Firefox's unfortunate choice to make a nonstandard string property magical.

**Comment by blesh** [comment]
I think we all would have been better off if people had just stuck to a few agreed upon conventions, honestly. But it's no big deal. It'll all be gone in a few years.

**Comment by ljharb** [comment]
I love your optimism :-) the first browser that didn't need the es5-shim appeared within the last year, 6 years after ES5 was standard.

**Comment by blesh** [comment]
Haha... Aw, 6 years isn't a big deal. Look how long it took Java to get generics or arrow functions, and that's just one platform.
